File: dom/src/test/java/xml/M0Test.java
package xml;
import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.Test;
import xml.m0.*;
import xml.m0.finalized.*;

public class M0Test {
    public interface TestTemplate<FX,FT,FD> extends Factory<FX,FT,FD> {
        default void test() {
            Tag<FX,FT,FD> tag = this.tag("html");
            assertTrue(tag.addChild(this.tag("head")));
            assertTrue(tag.addChild(this.tag("body")));

            assertTrue(tag.hasElem("body"));
            assertTrue(tag.notHasElem("script"));
            assertTrue(tag.deepClone().hasElem("head"));
            assertTrue(tag.deepClone().hasElem("body"));
            assertTrue(tag.deepClone().notHasElem("script"));

            XML<FX, FT, FD> text = this.text("test");
            assertTrue(text.hasElem("test"));
            assertTrue(text.notHasElem("html"));
            assertFalse(text.addChild(tag));
            assertFalse(text.addChild(text));
            text.deepClone().hasElem("test");
            text.deepClone().notHasElem("Test");

            Document<FX,FT,FD> singleRoot = this.singleRoot(tag);
            assertTrue(singleRoot.getRootTag().isPresent());
            assertSame(singleRoot.getRootTag().get(), tag);
            assertTrue(singleRoot.hasElem("body"));
            assertTrue(singleRoot.notHasElem("script"));
            assertTrue(singleRoot.deepClone().hasElem("head"));
            assertTrue(singleRoot.deepClone().hasElem("body"));
            assertTrue(singleRoot.deepClone().notHasElem("script"));
            assertTrue(singleRoot.sameRootElements(singleRoot.deepClone()));
            assertFalse(singleRoot.addChild(text));
            assertFalse(singleRoot.addChild(tag.deepClone()));
            assertFalse(singleRoot.sameRootElements(this.singleRoot(this.tag("head"))));
            assertFalse(singleRoot.sameRootElements(emptyDoc()));
        }

        Document<FX, FT, FD> emptyDoc();
    }

    private static class ActualTest implements TestTemplate<FXML, FTag, FDocument>, FFactory {
        @Override public xml.m0.finalized.FDocument emptyDoc() {
            return new xml.m0.finalized.FDocument() {
                @Override public boolean sameRootElements(Document<FXML, FTag, FDocument> doc) {
                    return doc.getRootTag().isEmpty();
                }

                @Override public Document<FXML, FTag, FDocument> deepClone() {
                    return emptyDoc();
                }

                @Override public boolean hasElem(String text) {
                    return false;
                }

                @Override public boolean addChild(XML<FXML, FTag, FDocument> child) {
                    return false;
                }
            };
        }
    }

    @Test
    public void testTest() { new ActualTest().test(); }
}

File: dom_scala/src/test/scala/xml/m0/M0Test.scala
package xml.m0

import org.scalatest.FunSuite

trait M0TestTemplate[T <: Types] extends FunSuite {
  val factory: Factory[T]
  val emptyDoc: Document[T]

  trait EmptyDoc extends Document[T] with Factory[T] {
    override def rootTag: Option[Tag[T]] = Option.empty
    def sameRootElements(doc: Document[T]): Boolean = doc.rootTag.isEmpty
    override def deepClone: Document[T] = emptyDoc
    def hasElem(text: String): Boolean = false
    def addChild(child: XML[T]): Boolean = false
  }

  test("M0") {
    import factory._
    val testTag = tag("html")
    assert(testTag.addChild(tag("head")))
    assert(testTag.addChild(tag("body")))
    assert(convert(testTag).eq(testTag))

    assert(testTag.hasElem("body"))
    assert(testTag.notHasElem("script"))
    assert(testTag.deepClone.hasElem("head"))
    assert(testTag.deepClone.hasElem("body"))
    assert(testTag.deepClone.notHasElem("script"))

    val testText = text("test")
    assert(testText.hasElem("test"))
    assert(testText.notHasElem("html"))
    assert(!testText.addChild(testTag))
    assert(!testText.addChild(testText))
    testText.deepClone.hasElem("test")
    testText.deepClone.notHasElem("Test")
    assert(convert(testText).eq(testText))

    val testSingleRoot = singleRoot(testTag)
    assert(testSingleRoot.rootTag.nonEmpty)
    assert(testSingleRoot.rootTag.get == testTag)
    assert(testSingleRoot.hasElem("body"))
    assert(testSingleRoot.notHasElem("script"))
    assert(testSingleRoot.deepClone.hasElem("head"))
    assert(testSingleRoot.deepClone.hasElem("body"))
    assert(testSingleRoot.deepClone.notHasElem("script"))
    assert(testSingleRoot.sameRootElements(testSingleRoot.deepClone))
    assert(!testSingleRoot.addChild(testText))
    assert(!testSingleRoot.addChild(testTag.deepClone))
    assert(!testSingleRoot.sameRootElements(singleRoot(tag("head"))))
    assert(!testSingleRoot.sameRootElements(emptyDoc))
    assert(convert(testSingleRoot).eq(testSingleRoot))
  }
}

class M0Test extends M0TestTemplate[finalized.FTypes] {
  val factory: Factory[finalized.FTypes] = new finalized.FFactory {}
  val emptyDoc = new EmptyDoc with finalized.FFactory with finalized.FDocument
}
